### DISCUSSION TITLE
V8

### DISCUSSION THREAD

**Discussion by hydroper** [airsdk/Adobe-Runtime-Support#2539] [Ideas]
V8 (the JavaScript engine used in Chromium, NodeJS and Deno) should be more ideal for executing JavaScript (specially TypeScript). Usually TypeScript is combined with bundling tools such as Webpack, Rollup, Parcel, Vite, Turbopack and more.

I'm thinking of translating a single untyped TypeScript bundle into AVM2 bytecode, but, besides it being possibly inefficient — not sure whether AVM2 does untyped code optimizations like property caching like V8 does — I'm not sure it'd be possible to install ANEs that way because, since it's an untyped JS bundle, things like classes are created as "constructor" functions with the `prototype` property assigned dynamically, and not as "AVM2 classes" (using the constant pool). The reason for that is inclusively because the class's "extends" clause is dynamic and can't be resolved ahead-of-time.

If Adobe AIR were to provide AIR developers with V8, I'm not sure how ANEs would be used, but it might work.

I'm not sure how difficult it might be to integrate V8 with the platform too... Just exploring!

### About AIR HTML content...

It looks like AIR supports HTML content, but it's only for _desktop_ devices.

- [Creating your first HTML-based AIR application with the AIR SDK](https://help.adobe.com/en_US/air/build/WS5b3ccc516d4fbf351e63e3d118666ade46-7ecc.html)

And the ANE question remains.

**Comment by ajwfrost** [comment]
Hi

I had wondered a while ago about whether it would help accessibility/take-up of AIR to switch to using V8 instead of the current ActionScript virtual machine, and then to have a way for people to compile their existing ActionScript functions into JavaScript. I think though, we would be more likely to look at the idea of a TypeScript to ABC compiler (like there is currently for C++ to ABC, with the 'Crossbridge' project) since switching out the code processing would be a massive effort, and trying to have multiple engines would likely be fraught with difficulties!

Plus don't forget V8 isn't allowed on iOS devices, 

The idea of an HTML-based AIR application is one I always found a little odd... but, that had been using a very old WebKit library that didn't then work on the new Apple silicon macOS platforms, so we have dropped that in favour of using the "StageWebView" component for providing HTML/JS capabilities. This just loads in whatever webview-type capability is provided by the operating system, which at least means it works across desktop and mobile devices, even though that also means that capabilities vary by platform.


Thanks for the discussion though .. I'd be interested in the background to your thoughts here i.e. why you're thinking it would be handy to have V8 provided in the AIR runtime?!

thanks

**Reply by hydroper** [reply]
Thanks for the reply, Andrew! I see... so V8 seems not a good option due to cross-platform support. Yeah, maybe the [TypeScript Compiler API](https://github.com/microsoft/TypeScript/wiki/Using-the-Compiler-API) could be used for translating TS to ABC. About why I thought of V8 being handy, it's simply because I don't know how to use the TS Compiler API at all. In the Microsoft TypeScript's GitHub wiki there's a bit of information on the Compiler API, but very little information, so you've to look at their compiler source.

**Reply by hydroper** [reply]
So, I think indeed TS to ABC would work well as TS resolves some symbols ahead of time, like the `extends` clause. Using the TSC API instead of taking a bundle from something like Webpack would simplify the TS codebase...

**Reply by hydroper** [reply]
Here's where the TypeScript's AST is defined at apparently: https://github.com/microsoft/TypeScript/blob/main/src/compiler/types.ts
- Ternary (aka. conditional expression): https://github.com/microsoft/TypeScript/blob/main/src/compiler/types.ts#L2725

In the project `wasm2swf` by https://github.com/brion there are few helpers for writting SWF, which can be useful for creating the TypeScript to ABC process:

- https://github.com/brion/wasm2swf/blob/master/src/swf.js

**Reply by ajwfrost** [reply]
That `wasm2swf` project looks fascinating! Could be very useful for this sort of scenario...

I had always thought that a potential workflow might involve exposing the ActionScript Flash/AIR APIs to TypeScript so that you could write some typescript that did something like
```
import { Rectangle } from "flash/geom/Rectangle.ts";
let rect : Rectangle = new Rectangle();
this.scrollRect = rect;
```
i.e. allow people to write AIR applications but using a different language (TypeScript)...

**Reply by hydroper** [reply]
I just don't like the idea of using WebAssembly or LLVM because it uses "pointers" and discards the garbage collector, so it'd be a waste of resources, I guess?

P.S.: compressing post here. Guess we should focus in TypeScript only.

**Reply by Fancy2209** [reply]
> i.e. allow people to write AIR applications but using a different language (TypeScript)...

Just like OpenFL with it's support for JS ES5, JS ES6 and TS? that'd sure be interesting!

**Reply by shaucky** [reply]
No no no… I have previously some benchmark Chromium/V8, Unity/CLR, and AIR/AVM2 and found that V8 may have 10 times the execution speed of AVM2 and 2 times that of CLR. So when Andrew said they considered embedding V8 into AIR, i was so expectant!
 
Compiling TypeScript into ABC and then executing it by AVM2 would be a very wrong choice, I think! This is likely to result in not being able to attract TypeScript developers as scheduled, lose more than gain…
 
-----
 
Sorry i didn't see the date of discussion

**Reply by ajwfrost** [reply]
It's still an ongoing discussion :-)

The two options really would be to convert from TypeScript into ActionScript Byte Code (or a form of this, like CrossBridge uses) vs embedding the V8 engine itself. I don't think I'd seen the above/earlier comment about the performance aspects; the downside of embedding V8 into AIR is a 8MB+ binary size increase. Although I guess it could be most feasible to just have two different runtimes, one that supports ABC and the other that supports JavaScript... (although a lot of the application bootstrap code is written in ActionScript so all of that could be an issue!).

Plus of course I think V8 is still forbidden on iOS? Or maybe it's okay if you don't use any other HTML rendering, it does seem to be supported: https://v8.dev/docs/cross-compile-ios

Interesting ideas here..!

**Reply by shaucky** [reply]
Addition: The benchmark test I mentioned was that one day last year, I and several chat group members wrote several basically consistent logic on their respective devices on Chromium/V8, Unity/CLR, and AIR/AVM2. The conclusion drawn is roughly like this, especially in the operation of integers, where their speed is approximately 10:5:1. V8 and AVM2 still have similar efficiency in floating-point operations.
